### ISSUE TITLE
Cannot delete from `__dict__` after `hasattr` test

### ISSUE THREAD

**Issue by AA-Turner** [astral-sh/ty#3081]
### Summary

Consider
```python
def f(obj: object, /) -> None:
    if hasattr(obj, '__dict__'):
        del obj.__dict__['f']  # Cannot delete subscript on object of type `object` with no `__delitem__` method (not-subscriptable)
    else:
        pass  # branch for slotted objects etc
```

Ty gives an error, but mypy, pyright, and pyrefly all pass.

Playground: https://play.ty.dev/37c236cc-cb6f-458d-8ab3-6e3678c0d6c9

If I add `and isinstance(obj.__dict__, dict)`, I get a different error: `` Method `__delitem__` of type `bound method Top[dict[Unknown, Unknown]].__delitem__(key: Never, /) -> None` cannot be called with key of type `Literal["f"]` on object of type `Top[dict[Unknown, Unknown]]` (invalid-argument-type) [Ln 3, Col 13] ``.

From the user's perspective I think the behaviour of the other three typecheckers is preferable here -- `__dict__` should always be a `dict` & so I should be able to delete from it.

If I'm reading the second error right it is saying that of all possible dicts, we cannot know that a key is present, and so it _could_ be an invalid argument. Would it be possible to distinguish this from the general case of `invalid-argument-type`? the un-minimised example deletes `self.__name__` from a cache within `__delete__`, so we 'know' that the key exists, but can't demonstrate this statically.

A

### Version

0.0.23 / playground f99b284c1

**Comment by sharkdp** [comment]
Thank you for reporting this. This is interesting.

For `hasattr(subject, "attr")` narrowing, we intersect the subject's type with a synthetic protocol `<Protocol with members 'attr'>`. In normal cases where `subject` is of type `Subject`, this means that the type of `subject` is now `Subject & <Protocol with members 'attr'>`. If you access attributes on that type, we look it up on both `Subject` and the synthetic protocol type, and then intersect those types of attributes. This means that you can still "see" the types of the actual attributes on `Subject`. For example:
```py
class Subject:
    attr: int

def _(subject: Subject):
    if hasattr(subject, "other_attr"):
        reveal_type(subject)  # Subject & <Protocol with members 'other_attr'>
        reveal_type(subject.attr)  # int
        reveal_type(subject.other_attr)  # object

    if hasattr(subject, "attr"):
        reveal_type(subject.attr)  # still int! (object & int = int)
```
https://play.ty.dev/83af0329-b3b0-4d7f-8b06-e9486ac45170


Your case is special because the `Subject` type itself is `object`. In that case, we simplify the intersection `object & <Protocol with members 'attr'>` to just `<Protocol with members 'attr'>`, because an intersection with "the universe" is a no-op.

However apparently, we then fail to see that the types of attributes on that synthetic protocol should still be intersected with the types of these attributes on the class `object`. So this looks like a clear bug to me.

**Comment by sharkdp** [comment]
> If I'm reading the second error right it is saying that of all possible dicts, we cannot know that a key is present, and so it _could_ be an invalid argument.

Exactly

> Would it be possible to distinguish this from the general case of `invalid-argument-type`?

Just to clarify, you're asking for a different diagnostic code / rule name for a case like this (minified a bit), so that you could suppress it more easily?

```py
def _(obj: object, /):
    if isinstance(obj, dict):
        del obj["some_key"]  # invalid-argument-type: Method `__delitem__` … can not be called …
```



> we 'know' that the key exists, but can't demonstrate this statically.

One possibility would be for ty to add some kind of key-based narrowing similar to `hasattr` narrowing. In the example above, you could additionally check for `… and "some_key" in obj`, and we could intersect with a synthetic `TypedDict` type that encodes the fact that a (non-required) item with key `"some_key"` exists.

**Comment by AA-Turner** [comment]
Thanks David!

> Just to clarify, you're asking for a different diagnostic code / rule name for a case like this (minified a bit), so that you could suppress it more easily?

Yes, I think this would be great. I'm working to integrate `ty` with a large legacy code base, and where possible I'm trying to avoid changes to runtime code (for now), at the expense of more suppressions. Here, I might like to globally ignore this diagnostic in `ty.toml`, but I don't want to globally ignore `invalid-argument-type`.

> One possibility would be for ty to add some kind of key-based narrowing similar to hasattr narrowing.

This also I think could be useful, but does have a slight TOCTOU problem, so I think having the flexibility to use either pattern would be great if possible.

A

**Comment by brent-at-aam** [comment]
I think there's just a larger problem in that `__dict__` is typed as an `object` even though it is a `dict[str, Any]`:

```python
def _(obj: object, /):
    if hasattr(obj, "__dict__"):
        d: dict[str, Any] = obj.__dict__ # Object of type `object` is not assignable to `dict[str, Any]`: Incompatible value of type `object`
```

`__annotations__` has the same behavior.

**Comment by carljm** [comment]
Let's use this issue to track the "we should assume attributes of `object`" problem.

I think having a different rule code for errors arising from top materializations can be tracked as part of #2221.

**Comment by carljm** [comment]
@brent-at-aam yes, that's the issue described above, where we need to treat attribute access on "<protocol with attribute 'whatever'>" as really an attribute access on `object & <Protocol with members 'whatever'>`, since `object` does actually have some typed attributes.
